homeassistant/components/esphome/media_player.py

```diff
@@ -98,10 +113,63 @@ async def async_play_media(
         media_id = async_process_play_media_url(self.hass, media_id)
         announcement = kwargs.get(ATTR_MEDIA_ANNOUNCE)
 
+        if (
+            self.supported_formats
+            and _is_url(media_id)
+            and (proxy_url := self._get_proxy_url(media_id, announcement is True))
+        ):
+            # Substitute proxy URL
+            media_id = proxy_url
+
         self._client.media_player_command(
             self._key, media_url=media_id, announcement=announcement
         )
 
+    def _get_proxy_url(self, url: str, announcement: bool) -> str | None:
+        """Get URL for ffmpeg proxy."""
+        if self.device_entry is None:
+            # Device id is required
+            return None
+
+        # Choose the first default or announcement supported format
+        format_to_use: MediaPlayerSupportedFormat | None = None
+        for supported_format in self.supported_formats:
+            if (format_to_use is None) and (
+                supported_format.purpose == MediaPlayerFormatPurpose.DEFAULT
+            ):
+                # First default format
+                format_to_use = supported_format
+            elif announcement and (
+                supported_format.purpose == MediaPlayerFormatPurpose.ANNOUNCEMENT
+            ):
+                # First announcement format
+                format_to_use = supported_format
+                break
+
+        if format_to_use is None:
+            # No format for conversion
+            return None
+
+        # Replace the media URL with a proxy URL pointing to Home
+        # Assistant. When requested, Home Assistant will use ffmpeg to
+        # convert the source URL to the supported format.
+        _LOGGER.debug("Proxying media url %s with format %s", url, format_to_use)
+        device_id = self.device_entry.id
+        media_format = format_to_use.format
+        convert_id = async_allow_proxy_url(
```

Comment:
If we are returning the URL, rename it:

```suggestion
        convert_id = async_create_proxy_url(
```